### DISCUSSION TITLE
why k_object_access_grant() execute fail

### DISCUSSION THREAD

**Discussion by frankey2009** [zephyrproject-rtos/zephyr#62187] [Q&A] [answered]
I define some vars for thread and timer:

K_THREAD_STACK_DEFINE(user_threadA_stack_area, STACKSIZE);
K_APP_BMEM(g_userspace_partition) static struct k_thread user_threadA_data;
struct k_timer g_user_k_timer;

next, I create thread for user mode:

tid = k_thread_create(&user_threadA_data, user_threadA_stack_area,
			K_THREAD_STACK_SIZEOF(user_threadA_stack_area),
			user_threadA_entry, NULL, NULL, NULL,
			PRIORITY, K_USER, K_FOREVER);

 k_thread_name_set(&user_threadA_data, "user_thread_a");
 k_mem_domain_add_thread(&g_app_domain, tid);
 k_object_access_grant(&g_user_k_timer, tid);
 k_thread_start(tid);

when make success, I found no kernel thread(K_OBJ_THREAD) for this userspace thread,when I run this code,
I found k_object_access_grant() not give timer access authority to this thread, so I debug this function,I found 
![image](https://github.com/zephyrproject-rtos/zephyr/assets/18458784/1fd70a0c-ddfc-472f-9278-34d252ab95d4)

void z_thread_perms_set(struct z_object *ko, struct k_thread *thread)
{
	int index = thread_index_get(thread);

	if (index != -1) {
		sys_bitfield_set_bit((mem_addr_t)&ko->perms, index);
	}
}

in this function, not fine this userspace thread object !!!

so my question as follows:
1.why k_object_access_grant()  API not give fail return tips?
2.why this userspace thread not find in kernel object,that is the cause for k_object_access_grant() execute fail?
3.what is the right method for this scene,give K_OBJ_TIMER kernel object access authority to the user mode thread?

**Comment by mcscholtz** [comment, marked as answer]
I am having the same problem with Zephyr 3.4. I got it to work for regular kernel objects such as mutex's by dynamically allocating them using `k_object_alloc` 

e.g
```
struct k_mutex *mutex = k_object_alloc(K_OBJ_MUTEX);
k_mutex_init(mutex);
k_object_access_grant(mutex, tid);
```
Should work for the timer also, you can also try initializing it first and see if that works.

However I cannot get it to work for any device drivers which are all static objects. It behaves like you described above and I cannot figure out why. Only making them public objects works which is not what I want to do.

**Reply by jukkar** [reply]
See for example
`tests/kernel/mem_protect/mem_protect/src/kobject.c` and `test_kobject_access_grant` test how it sets the access to static mutex.

**Reply by mcscholtz** [reply]
Maybe it  works for a mutex, does not work for devices

**Reply by dcpleung** [reply]
That is by design. A user thread should not have directly access to devices. You will need to implement syscalls where it will run in kernel mode to access the devices.

**Reply by mcscholtz** [reply]
Yeah that makes a lot of sense now that I think about it. Thanks
